NPE in initiateHoverRequest (this.cancellationSupport is null)

Hover can fail with an NPE at

https://github.com/redhat-developer/intellij-quarkus/blob/bd549b44cdfa9dace1714bdc24ba9065b04c8986/src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/hover/LSPTextHover.java#L204

```
2023-08-11 10:54:47,172 [  27144]   WARN - c.r.d.i.l.o.h.LSPTextHover - java.lang.NullPointerException: Cannot invoke "com.redhat.devtools.intellij.lsp4ij.internal.CancellationSupport.execute(java.util.concurrent.CompletableFuture)" because "this.cancellationSupport" is null
java.util.concurrent.ExecutionException: java.lang.NullPointerException: Cannot invoke "com.redhat.devtools.intellij.lsp4ij.internal.CancellationSupport.execute(java.util.concurrent.CompletableFuture)" because "this.cancellationSupport" is null
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2096)
	at com.redhat.devtools.intellij.lsp4ij.operations.hover.LSPTextHover.generateDoc(LSPTextHover.java:123)
	at com.intellij.lang.documentation.CompositeDocumentationProvider.generateDoc(CompositeDocumentationProvider.java:135)
	at com.intellij.lang.documentation.psi.PsiElementDocumentationTarget.localDocHtml(PsiElementDocumentationTarget.kt:102)
	at com.intellij.lang.documentation.psi.PsiElementDocumentationTarget.localDoc(PsiElementDocumentationTarget.kt:88)
	at com.intellij.lang.documentation.psi.PsiElementDocumentationTarget.computeDocumentation(PsiElementDocumentationTarget.kt:75)
	at com.intellij.lang.documentation.impl.ImplKt$computeDocumentation$2$documentationResult$1.invoke(impl.kt:31)
	at com.intellij.lang.documentation.impl.ImplKt$computeDocumentation$2$documentationResult$1.invoke(impl.kt)
	at com.intellij.openapi.application.rw.InternalReadAction.insideReadAction(InternalReadAction.kt:96)
	at com.intellij.openapi.application.rw.InternalReadAction.access$insideReadAction(InternalReadAction.kt:13)
	at com.intellij.openapi.application.rw.InternalReadAction$tryReadCancellable$1.invoke(InternalReadAction.kt:81)
	at com.intellij.openapi.application.rw.InternalReadAction$tryReadCancellable$1.invoke(InternalReadAction.kt:13)
	at com.intellij.openapi.progress.CancellationKt$sam$com_intellij_openapi_util_ThrowableComputable$0.compute(cancellation.kt)
	at com.intellij.openapi.progress.Cancellation.withJob(Cancellation.java:60)
	at com.intellij.openapi.progress.CancellationKt.withJob(cancellation.kt:14)
	at com.intellij.openapi.progress.CancellationKt.executeWithJobAndCompleteIt(cancellation.kt:111)
	at com.intellij.openapi.application.rw.CancellableReadActionKt$cancellableReadActionInternal$2$1.run(cancellableReadAction.kt:33)
	at com.intellij.openapi.application.impl.ApplicationImpl.tryRunReadAction(ApplicationImpl.java:1154)
	at com.intellij.openapi.application.rw.CancellableReadActionKt$cancellableReadActionInternal$2.run(cancellableReadAction.kt:31)
	at com.intellij.openapi.progress.util.ProgressIndicatorUtils.runActionAndCancelBeforeWrite(ProgressIndicatorUtils.java:158)
	at com.intellij.openapi.application.rw.CancellableReadActionKt.cancellableReadActionInternal(cancellableReadAction.kt:29)
	at com.intellij.openapi.application.rw.InternalReadAction.tryReadCancellable(InternalReadAction.kt:80)
	at com.intellij.openapi.application.rw.InternalReadAction.access$tryReadCancellable(InternalReadAction.kt:13)
	at com.intellij.openapi.application.rw.InternalReadAction$tryReadAction$2.invoke(InternalReadAction.kt:66)
	at com.intellij.openapi.application.rw.InternalReadAction$tryReadAction$2.invoke(InternalReadAction.kt:13)
	at com.intellij.openapi.progress.CancellationKt$sam$com_intellij_openapi_util_ThrowableComputable$0.compute(cancellation.kt)
	at com.intellij.openapi.progress.Cancellation.withJob(Cancellation.java:60)
	at com.intellij.openapi.progress.CancellationKt.withJob(cancellation.kt:14)
	at com.intellij.openapi.progress.CoroutinesKt.blockingContext(coroutines.kt:138)
	at com.intellij.openapi.application.rw.InternalReadAction.tryReadAction(InternalReadAction.kt:61)
	at com.intellij.openapi.application.rw.InternalReadAction.readLoop(InternalReadAction.kt:53)
	at com.intellij.openapi.application.rw.InternalReadAction.access$readLoop(InternalReadAction.kt:13)
	at com.intellij.openapi.application.rw.InternalReadAction$runReadAction$4.invokeSuspend(InternalReadAction.kt:33)
	at com.intellij.openapi.application.rw.InternalReadAction$runReadAction$4.invoke(InternalReadAction.kt)
	at kotlinx.coroutines.intrinsics.UndispatchedKt.startUndispatchedOrReturn(Undispatched.kt:89)
	at kotlinx.coroutines.CoroutineScopeKt.coroutineScope(CoroutineScope.kt:264)
	at com.intellij.openapi.application.rw.InternalReadAction.runReadAction(InternalReadAction.kt:32)
	at com.intellij.openapi.application.rw.PlatformReadActionSupport.executeReadAction(PlatformReadActionSupport.kt:24)
	at com.intellij.openapi.application.CoroutinesKt.constrainedReadAction(coroutines.kt:47)
	at com.intellij.openapi.application.CoroutinesKt.readAction(coroutines.kt:17)
	at com.intellij.lang.documentation.impl.ImplKt$computeDocumentation$2.invokeSuspend(impl.kt:30)
	at com.intellij.lang.documentation.impl.ImplKt$computeDocumentation$2.invoke(impl.kt)
	at kotlinx.coroutines.intrinsics.UndispatchedKt.startUndispatchedOrReturn(Undispatched.kt:89)
	at kotlinx.coroutines.BuildersKt__Builders_commonKt.withContext(Builders.common.kt:165)
	at kotlinx.coroutines.BuildersKt.withContext(Unknown Source)
	at com.intellij.lang.documentation.impl.ImplKt.computeDocumentation(impl.kt:29)
	at com.intellij.lang.documentation.ide.impl.DocumentationPage.loadPage(DocumentationPage.kt:32)
	at com.intellij.lang.documentation.ide.impl.DocumentationBrowser.handleBrowserRequest(DocumentationBrowser.kt:91)
	at com.intellij.lang.documentation.ide.impl.DocumentationBrowser.access$handleBrowserRequest(DocumentationBrowser.kt:24)
	at com.intellij.lang.documentation.ide.impl.DocumentationBrowser$1$1.invokeSuspend(DocumentationBrowser.kt:51)
	at com.intellij.lang.documentation.ide.impl.DocumentationBrowser$1$1.invoke(DocumentationBrowser.kt)
	at kotlinx.coroutines.flow.FlowKt__MergeKt$mapLatest$1.invokeSuspend(Merge.kt:217)
	at kotlinx.coroutines.flow.FlowKt__MergeKt$mapLatest$1.invoke(Merge.kt)
	at kotlinx.coroutines.flow.FlowKt__MergeKt$mapLatest$1.invoke(Merge.kt)
	at kotlinx.coroutines.flow.internal.ChannelFlowTransformLatest$flowCollect$3$1$2.invokeSuspend(Merge.kt:34)
	at kotlinx.coroutines.flow.internal.ChannelFlowTransformLatest$flowCollect$3$1$2.invoke(Merge.kt)
	at kotlinx.coroutines.flow.internal.ChannelFlowTransformLatest$flowCollect$3$1$2.invoke(Merge.kt)
	at kotlinx.coroutines.intrinsics.UndispatchedKt.startCoroutineUndispatched(Undispatched.kt:55)
	at kotlinx.coroutines.CoroutineStart.invoke(CoroutineStart.kt:112)
	at kotlinx.coroutines.AbstractCoroutine.start(AbstractCoroutine.kt:126)
	at kotlinx.coroutines.BuildersKt__Builders_commonKt.launch(Builders.common.kt:56)
	at kotlinx.coroutines.BuildersKt.launch(Unknown Source)
	at kotlinx.coroutines.BuildersKt__Builders_commonKt.launch$default(Builders.common.kt:47)
	at kotlinx.coroutines.BuildersKt.launch$default(Unknown Source)
	at kotlinx.coroutines.flow.internal.ChannelFlowTransformLatest$flowCollect$3$invokeSuspend$$inlined$collect$1.emit(Collect.kt:140)
	at kotlinx.coroutines.flow.SharedFlowImpl.collect(SharedFlow.kt:351)
	at kotlinx.coroutines.flow.internal.ChannelFlowTransformLatest$flowCollect$3.invokeSuspend(Merge.kt:101)
	at kotlinx.coroutines.flow.internal.ChannelFlowTransformLatest$flowCollect$3.invoke(Merge.kt)
	at kotlinx.coroutines.flow.internal.ChannelFlowTransformLatest$flowCollect$3.invoke(Merge.kt)
	at kotlinx.coroutines.intrinsics.UndispatchedKt.startUndispatchedOrReturn(Undispatched.kt:89)
	at kotlinx.coroutines.flow.internal.FlowCoroutineKt.flowScope(FlowCoroutine.kt:33)
	at kotlinx.coroutines.flow.internal.ChannelFlowTransformLatest.flowCollect(Merge.kt:25)
	at kotlinx.coroutines.flow.internal.ChannelFlowOperator.collectTo$suspendImpl(ChannelFlow.kt:157)
	at kotlinx.coroutines.flow.internal.ChannelFlowOperator.collectTo(ChannelFlow.kt)
	at kotlinx.coroutines.flow.internal.ChannelFlow$collectToFun$1.invokeSuspend(ChannelFlow.kt:60)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
Caused by: java.lang.NullPointerException: Cannot invoke "com.redhat.devtools.intellij.lsp4ij.internal.CancellationSupport.execute(java.util.concurrent.CompletableFuture)" because "this.cancellationSupport" is null
	at com.redhat.devtools.intellij.lsp4ij.operations.hover.LSPTextHover.lambda$initiateHoverRequest$2(LSPTextHover.java:204)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1625)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
	at com.redhat.devtools.intellij.lsp4ij.operations.hover.LSPTextHover.lambda$initiateHoverRequest$3(LSPTextHover.java:208)
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:646)
	at java.base/java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:483)
Caused by: java.lang.NullPointerException: Cannot invoke "com.redhat.devtools.intellij.lsp4ij.internal.CancellationSupport.execute(java.util.concurrent.CompletableFuture)" because "this.cancellationSupport" is null

	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
```

--- a/src/main/java/com/redhat/devtools/intellij/lsp4ij/LSPVirtualFileWrapper.java
+++ b/src/main/java/com/redhat/devtools/intellij/lsp4ij/LSPVirtualFileWrapper.java
@@ -15,12 +15,15 @@
 package com.redhat.devtools.intellij.lsp4ij;
 
 import com.intellij.openapi.Disposable;
+import com.intellij.openapi.editor.Editor;
 import com.intellij.openapi.module.Module;
 import com.intellij.openapi.util.Disposer;
 import com.intellij.openapi.util.Key;
 import com.intellij.openapi.vfs.VirtualFile;
+import com.intellij.psi.PsiElement;
 import com.redhat.devtools.intellij.lsp4ij.operations.diagnostics.LSPDiagnosticsForServer;
 import com.redhat.devtools.intellij.lsp4ij.operations.documentLink.LSPDocumentLinkForServer;
+import com.redhat.devtools.intellij.lsp4ij.operations.hover.LSPTextHoverForFile;
 import org.eclipse.lsp4j.Diagnostic;
 import org.eclipse.lsp4j.DocumentLink;
 
@@ -45,10 +48,12 @@ public class LSPVirtualFileWrapper implements Disposable {
     private final VirtualFile file;
 
     private final Map<LanguageServerWrapper, LSPVirtualFileData /* current LSP data (diagnostics, documentLink, etc) */> dataPerServer;
+    private final LSPTextHoverForFile hover;
 
     LSPVirtualFileWrapper(VirtualFile file) {
         this.file = file;
         this.dataPerServer = new HashMap<>();
+        this.hover = new LSPTextHoverForFile();
         Module project = LSPIJUtils.getProject(file);
         if (project != null) {
             Disposer.register(project, this);
@@ -101,6 +106,11 @@ public class LSPVirtualFileWrapper implements Disposable {
      */
     public Collection<LSPDocumentLinkForServer> getAllDocumentLink() {
         return getData(LSPVirtualFileData::getDocumentLinkForServer);
+    }
+
+
+    public String getHoverContent(PsiElement element, int targetOffset, Editor editor) {
+        return hover.getHoverContent(element, targetOffset, editor);
     }
 
     // ------------------------ Other methods
@@ -176,4 +186,5 @@ public class LSPVirtualFileWrapper implements Disposable {
             wrapper.dispose();
         }
     }
+
 }

--- a/src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/hover/LSPTextHover.java
+++ b/src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/hover/LSPTextHover.java
@@ -15,11 +15,14 @@ import com.intellij.lang.documentation.ExternalDocumentationHandler;
 import com.intellij.openapi.editor.Document;
 import com.intellij.openapi.editor.Editor;
 import com.intellij.openapi.fileEditor.FileEditorManager;
+import com.intellij.openapi.project.Project;
 import com.intellij.openapi.vfs.VirtualFile;
 import com.intellij.psi.PsiDocumentManager;
 import com.intellij.psi.PsiElement;
+import com.intellij.psi.PsiFile;
 import com.intellij.psi.PsiManager;
 import com.redhat.devtools.intellij.lsp4ij.LSPIJUtils;
+import com.redhat.devtools.intellij.lsp4ij.LSPVirtualFileWrapper;
 import com.redhat.devtools.intellij.lsp4ij.LanguageServiceAccessor;
 import com.redhat.devtools.intellij.lsp4ij.internal.CancellationSupport;
 import com.vladsch.flexmark.html.HtmlRenderer;
@@ -44,62 +47,11 @@ import java.util.concurrent.TimeoutException;
 import java.util.function.Predicate;
 import java.util.stream.Collectors;
 
+/**
+ * LSP testDocument/hover support.
+ *
+ */
 public class LSPTextHover extends DocumentationProviderEx implements ExternalDocumentationHandler {
-    private static final Logger LOGGER = LoggerFactory.getLogger(LSPTextHover.class);
-
-    private static final String HEAD = "<head>"; //$NON-NLS-1$
-
-    private static final Parser PARSER = Parser.builder().build();
-    private static final HtmlRenderer RENDERER = HtmlRenderer.builder().build();
-
-    private PsiElement lastElement;
-    private int lastOffset = -1;
-    private CompletableFuture<List<Hover>> lspRequest;
-    private CancellationSupport cancellationSupport;
-
-    public LSPTextHover() {
-        LOGGER.info("LSPTextHover");
-    }
-
-    public static String styleHtml(Editor editor, String html) {
-        if (html == null || html.isEmpty()) {
-            return html;
-        }
-        Color background = editor.getColorsScheme().getDefaultBackground();
-        Color foreground = editor.getColorsScheme().getDefaultForeground();
-        // put CSS styling to match Eclipse style
-        String style = "<html><head><style TYPE='text/css'>html { " + //$NON-NLS-1$
-                (background != null ? "background-color: " + toHTMLrgb(background) + "; " : "") + //$NON-NLS-1$ //$NON-NLS-2$ //$NON-NLS-3$
-                (foreground != null ? "color: " + toHTMLrgb(foreground) + "; " : "") + //$NON-NLS-1$ //$NON-NLS-2$ //$NON-NLS-3$
-                " }</style></head><body>"; //$NON-NLS-1$
-
-        /*int headIndex = html.indexOf(HEAD);
-        StringBuilder builder = new StringBuilder(html.length() + style.length());
-        builder.append(html.substring(0, headIndex + HEAD.length()));
-        builder.append(style);
-        builder.append(html.substring(headIndex + HEAD.length()));
-        return builder.toString();*/
-        StringBuilder builder = new StringBuilder(style);
-        builder.append(html).append("</body></html>");
-        return builder.toString();
-    }
-
-    private static String toHTMLrgb(Color rgb) {
-        StringBuilder builder = new StringBuilder(7);
-        builder.append('#');
-        appendAsHexString(builder, rgb.getRed());
-        appendAsHexString(builder, rgb.getGreen());
-        appendAsHexString(builder, rgb.getBlue());
-        return builder.toString();
-    }
-
-    private static void appendAsHexString(StringBuilder buffer, int intValue) {
-        String hexValue = Integer.toHexString(intValue);
-        if (hexValue.length() == 1) {
-            buffer.append('0');
-        }
-        buffer.append(hexValue);
-    }
 
     @Nullable
     @Override
@@ -110,6 +62,10 @@ public class LSPTextHover extends DocumentationProviderEx implements ExternalDoc
     @Nullable
     @Override
     public String generateDoc(PsiElement element, @Nullable PsiElement originalElement) {
+        Project project = element.getProject();
+        if (project.isDisposed()) {
+            return null;
+        }
         if (originalElement == null || !Objects.equals(element.getContainingFile(), originalElement.getContainingFile())) {
             return null;
         }
@@ -117,100 +73,16 @@ public class LSPTextHover extends DocumentationProviderEx implements ExternalDoc
         if (editor == null) {
             return null;
         }
+
+        VirtualFile file = originalElement.getContainingFile().getVirtualFile();
         int targetOffset = getTargetOffset(originalElement);
-        initiateHoverRequest(element, targetOffset);
-        try {
-            String result = lspRequest.get(500, TimeUnit.MILLISECONDS).stream()
-                    .filter(Objects::nonNull)
-                    .map(LSPTextHover::getHoverString)
-                    .filter(Objects::nonNull)
-                    .collect(Collectors.joining("\n\n")) //$NON-NLS-1$
-                    .trim();
-            if (!result.isEmpty()) {
-                return styleHtml(editor, RENDERER.render(PARSER.parse(result)));
-            }
-        } catch (ExecutionException | TimeoutException e) {
-            LOGGER.warn(e.getLocalizedMessage(), e);
-        } catch (InterruptedException e) {
-            LOGGER.warn(e.getLocalizedMessage(), e);
-            Thread.currentThread().interrupt();
-        } finally {
-            this.cancellationSupport = null;
-        }
-        return null;
+        return LSPVirtualFileWrapper.getLSPVirtualFileWrapper(file).getHoverContent(element, targetOffset, editor);
     }
 
-    private int getTargetOffset(PsiElement originalElement) {
+    private static int getTargetOffset(PsiElement originalElement) {
         int startOffset = originalElement.getTextOffset();
         int textLength = originalElement.getTextLength();
         return startOffset + textLength / 2;
-    }
-
-    private static @Nullable String getHoverString(Hover hover) {
-        Either<List<Either<String, MarkedString>>, MarkupContent> hoverContent = hover.getContents();
-        if (hoverContent.isLeft()) {
-            List<Either<String, MarkedString>> contents = hoverContent.getLeft();
-            if (contents == null || contents.isEmpty()) {
-                return null;
-            }
-            return contents.stream().map(content -> {
-                if (content.isLeft()) {
-                    return content.getLeft();
-                } else if (content.isRight()) {
-                    MarkedString markedString = content.getRight();
-                    // TODO this won't work fully until markup parser will support syntax
-                    // highlighting but will help display
-                    // strings with language tags, e.g. without it things after <?php tag aren't
-                    // displayed
-                    if (markedString.getLanguage() != null && !markedString.getLanguage().isEmpty()) {
-                        return String.format("```%s%n%s%n```", markedString.getLanguage(), markedString.getValue()); //$NON-NLS-1$
-                    } else {
-                        return markedString.getValue();
-                    }
-                } else {
-                    return ""; //$NON-NLS-1$
-                }
-            }).filter(((Predicate<String>) String::isEmpty).negate()).collect(Collectors.joining("\n\n")); //$NON-NLS-1$ )
-        } else {
-            return hoverContent.getRight().getValue();
-        }
-    }
-
-
-    /**
-     * Initialize hover requests with hover (if available) and codelens (if
-     * available).
-     *
-     * @param element the PSI element.
-     * @param offset  the target offset.
-     */
-    private void initiateHoverRequest(PsiElement element, int offset) {
-        if (this.cancellationSupport != null) {
-            this.cancellationSupport.cancel();
-            this.cancellationSupport = null;
-        }
-        PsiDocumentManager manager = PsiDocumentManager.getInstance(element.getProject());
-        final Document document = manager.getDocument(element.getContainingFile());
-        if (offset != -1 && (this.lspRequest == null || !element.equals(this.lastElement) || offset != this.lastOffset)) {
-            this.lastElement = element;
-            this.lastOffset = offset;
-            this.cancellationSupport = new CancellationSupport();
-            this.lspRequest = LanguageServiceAccessor.getInstance(element.getProject())
-                    .getLanguageServers(document, capabilities -> isHoverCapable(capabilities))
-                    .thenApplyAsync(languageServers -> // Async is very important here, otherwise the LS Client thread is in
-                            // deadlock and doesn't read bytes from LS
-                            languageServers.stream()
-                                    .map(languageServer ->
-                                            cancellationSupport.execute(
-                                                            languageServer.getServer().getTextDocumentService()
-                                                                    .hover(LSPIJUtils.toHoverParams(offset, document)))
-                                                    .join()
-                                    ).filter(Objects::nonNull).collect(Collectors.toList()));
-        }
-    }
-
-    private boolean isHoverCapable(ServerCapabilities capabilities) {
-        return (capabilities.getHoverProvider().isLeft() && capabilities.getHoverProvider().getLeft()) || capabilities.getHoverProvider().isRight();
     }
 
     @Nullable

--- a/src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/hover/LSPTextHoverForFile.java
+++ b/src/main/java/com/redhat/devtools/intellij/lsp4ij/operations/hover/LSPTextHoverForFile.java
@@ -1,0 +1,188 @@
+/*******************************************************************************
+ * Copyright (c) 2023 Red Hat, Inc.
+ * Distributed under license by Red Hat, Inc. All rights reserved.
+ * This program is made available under the terms of the
+ * Eclipse Public License v2.0 which accompanies this distribution,
+ * and is available at https://www.eclipse.org/legal/epl-v20.html
+ *
+ * Contributors:
+ * Red Hat, Inc. - initial API and implementation
+ ******************************************************************************/
+package com.redhat.devtools.intellij.lsp4ij.operations.hover;
+
+import com.intellij.openapi.editor.Document;
+import com.intellij.openapi.editor.Editor;
+import com.intellij.psi.PsiDocumentManager;
+import com.intellij.psi.PsiElement;
+import com.redhat.devtools.intellij.lsp4ij.LSPIJUtils;
+import com.redhat.devtools.intellij.lsp4ij.LanguageServiceAccessor;
+import com.redhat.devtools.intellij.lsp4ij.internal.CancellationSupport;
+import com.vladsch.flexmark.html.HtmlRenderer;
+import com.vladsch.flexmark.parser.Parser;
+import org.eclipse.lsp4j.Hover;
+import org.eclipse.lsp4j.MarkedString;
+import org.eclipse.lsp4j.MarkupContent;
+import org.eclipse.lsp4j.ServerCapabilities;
+import org.eclipse.lsp4j.jsonrpc.messages.Either;
+import org.jetbrains.annotations.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.awt.*;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+/**
+ * LSP textDocument/hover support for a given file.
+ */
+public class LSPTextHoverForFile {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(LSPTextHoverForFile.class);
+    private static final Parser PARSER = Parser.builder().build();
+    private static final HtmlRenderer RENDERER = HtmlRenderer.builder().build();
+
+    private PsiElement lastElement;
+    private int lastOffset = -1;
+    private CompletableFuture<List<Hover>> lspRequest;
+    private CancellationSupport previousCancellationSupport;
+
+    public String getHoverContent(PsiElement element, int targetOffset, Editor editor) {
+        initiateHoverRequest(element, targetOffset);
+        try {
+            String result = lspRequest.get(500, TimeUnit.MILLISECONDS).stream()
+                    .filter(Objects::nonNull)
+                    .map(LSPTextHoverForFile::getHoverString)
+                    .filter(Objects::nonNull)
+                    .collect(Collectors.joining("\n\n")) //$NON-NLS-1$
+                    .trim();
+            if (!result.isEmpty()) {
+                return styleHtml(editor, RENDERER.render(PARSER.parse(result)));
+            }
+            // The LSP hover request are finished, don't need to cancel the previous LSP requests.
+            previousCancellationSupport = null;
+        } catch (ExecutionException | TimeoutException e) {
+            LOGGER.warn(e.getLocalizedMessage(), e);
+        } catch (InterruptedException e) {
+            LOGGER.warn(e.getLocalizedMessage(), e);
+            Thread.currentThread().interrupt();
+        }
+        return null;
+    }
+
+
+    /**
+     * Initialize hover requests with hover (if available).
+     *
+     * @param element the PSI element.
+     * @param offset  the target offset.
+     */
+
+    private void initiateHoverRequest(PsiElement element, int offset) {
+        if (this.previousCancellationSupport != null) {
+            // The prvious LSP hover request is not finished,cancel it
+            this.previousCancellationSupport.cancel();
+        }
+        PsiDocumentManager manager = PsiDocumentManager.getInstance(element.getProject());
+        final Document document = manager.getDocument(element.getContainingFile());
+        if (offset != -1 && (this.lspRequest == null || !element.equals(this.lastElement) || offset != this.lastOffset)) {
+            this.lastElement = element;
+            this.lastOffset = offset;
+            CancellationSupport cancellationSupport = new CancellationSupport();
+            this.lspRequest = LanguageServiceAccessor.getInstance(element.getProject())
+                    .getLanguageServers(document, capabilities -> isHoverCapable(capabilities))
+                    .thenApplyAsync(languageServers -> // Async is very important here, otherwise the LS Client thread is in
+                            // deadlock and doesn't read bytes from LS
+                            languageServers.stream()
+                                    .map(languageServer ->
+                                            cancellationSupport.execute(
+                                                            languageServer.getServer().getTextDocumentService()
+                                                                    .hover(LSPIJUtils.toHoverParams(offset, document)))
+                                                    .join()
+                                    ).filter(Objects::nonNull).collect(Collectors.toList()));
+            // store the current cancel support as previous
+            previousCancellationSupport = cancellationSupport;
+        }
+    }
+
+    private static @Nullable String getHoverString(Hover hover) {
+        Either<List<Either<String, MarkedString>>, MarkupContent> hoverContent = hover.getContents();
+        if (hoverContent.isLeft()) {
+            List<Either<String, MarkedString>> contents = hoverContent.getLeft();
+            if (contents == null || contents.isEmpty()) {
+                return null;
+            }
+            return contents.stream().map(content -> {
+                if (content.isLeft()) {
+                    return content.getLeft();
+                } else if (content.isRight()) {
+                    MarkedString markedString = content.getRight();
+                    // TODO this won't work fully until markup parser will support syntax
+                    // highlighting but will help display
+                    // strings with language tags, e.g. without it things after <?php tag aren't
+                    // displayed
+                    if (markedString.getLanguage() != null && !markedString.getLanguage().isEmpty()) {
+                        return String.format("```%s%n%s%n```", markedString.getLanguage(), markedString.getValue()); //$NON-NLS-1$
+                    } else {
+                        return markedString.getValue();
+                    }
+                } else {
+                    return ""; //$NON-NLS-1$
+                }
+            }).filter(((Predicate<String>) String::isEmpty).negate()).collect(Collectors.joining("\n\n")); //$NON-NLS-1$ )
+        } else {
+            return hoverContent.getRight().getValue();
+        }
+    }
+
+    private static boolean isHoverCapable(ServerCapabilities capabilities) {
+        return (capabilities.getHoverProvider().isLeft() && capabilities.getHoverProvider().getLeft()) || capabilities.getHoverProvider().isRight();
+    }
+
+
+    public static String styleHtml(Editor editor, String html) {
+        if (html == null || html.isEmpty()) {
+            return html;
+        }
+        Color background = editor.getColorsScheme().getDefaultBackground();
+        Color foreground = editor.getColorsScheme().getDefaultForeground();
+        // put CSS styling to match Eclipse style
+        String style = "<html><head><style TYPE='text/css'>html { " + //$NON-NLS-1$
+                (background != null ? "background-color: " + toHTMLrgb(background) + "; " : "") + //$NON-NLS-1$ //$NON-NLS-2$ //$NON-NLS-3$
+                (foreground != null ? "color: " + toHTMLrgb(foreground) + "; " : "") + //$NON-NLS-1$ //$NON-NLS-2$ //$NON-NLS-3$
+                " }</style></head><body>"; //$NON-NLS-1$
+
+        /*int headIndex = html.indexOf(HEAD);
+        StringBuilder builder = new StringBuilder(html.length() + style.length());
+        builder.append(html.substring(0, headIndex + HEAD.length()));
+        builder.append(style);
+        builder.append(html.substring(headIndex + HEAD.length()));
+        return builder.toString();*/
+        StringBuilder builder = new StringBuilder(style);
+        builder.append(html).append("</body></html>");
+        return builder.toString();
+    }
+
+    private static String toHTMLrgb(Color rgb) {
+        StringBuilder builder = new StringBuilder(7);
+        builder.append('#');
+        appendAsHexString(builder, rgb.getRed());
+        appendAsHexString(builder, rgb.getGreen());
+        appendAsHexString(builder, rgb.getBlue());
+        return builder.toString();
+    }
+
+    private static void appendAsHexString(StringBuilder buffer, int intValue) {
+        String hexValue = Integer.toHexString(intValue);
+        if (hexValue.length() == 1) {
+            buffer.append('0');
+        }
+        buffer.append(hexValue);
+    }
+
+}

